src/bencode/bencode_parser.cpp

```diff
@@ -8,21 +8,19 @@ std::string BencodeParser::encode(const BencodeValue& value) {
   return value.to_string();
 }
 
-std::unique_ptr<BencodeValue> BencodeParser::decode(
-    const std::string& decoded) {
+auto BencodeParser::decode(const std::string& decoded) -> BencodeResult{
   _tokens = decoded;
   _index = 0;  // Reset the index
   auto r = decode();
-  if(_index != _tokens.size()) {
-    throw std::invalid_argument(
-        "BencodeParser::decode(std::string decoded): invalid decoded string");
+  if(r.valid() && _index != _tokens.size()) {
+    // If the result is not an error and the string was not fully parsed
+    return BencodeResult::ERROR(BencodeParserError::InvalidString);
   }
   return r;
 }
 
 // Given a vector of tokens, returns a BencodeValue object
-std::unique_ptr<BencodeValue> BencodeParser::decode() {
-  std::regex reg_string("^\\d+:$");
+auto BencodeParser::decode() -> BencodeResult{
```

Comment:
Also, I'm not really getting the need for this syntax (which is inconsistent with `bencode_parser.hpp` in this specific case). As far as I can see, it has [some specific use cases](https://stackoverflow.com/a/22515589) but none applies here. I'm not explicitly against it though, just wondering why it's there.